automated_analysis.py

```diff
@@ -109,46 +131,13 @@
         for row in engagement_counts.values():
             writer.writerow(row)
 
-    log.info("Computing the participation frequencies...")
-    repeat_participations = OrderedDict()
-    for i in range(1, len(PipelineConfiguration.RQA_CODING_PLANS) + 1):
-        repeat_participations[i] = {
-            "Number of Episodes Participated In": i,
-            "Number of Participants with Opt-Ins": 0,
-            "% of Participants with Opt-Ins": None
-        }
-
-    # Compute the number of individuals who participated each possible number of times, from 1 to <number of RQAs>
-    # An individual is considered to have participated if they sent a message and didn't opt-out, regardless of the
-    # relevance of any of their messages.
-    for ind in individuals:
-        if AnalysisUtils.withdrew_consent(ind, CONSENT_WITHDRAWN_KEY):
-            continue
-
-        weeks_participated = 0
-        for plan in PipelineConfiguration.RQA_CODING_PLANS:
-            if AnalysisUtils.opt_in(ind, CONSENT_WITHDRAWN_KEY, plan):
-                weeks_participated += 1
-        assert weeks_participated != 0, f"Found individual '{ind['uid']}' with no participation in any week"
-        repeat_participations[weeks_participated]["Number of Participants with Opt-Ins"] += 1
-
-    # Compute the percentage of individuals who participated each possible number of times.
-    # Percentages are computed out of the total number of participants who opted-in.
-    total_participants = len(AnalysisUtils.filter_opt_ins(
-        individuals, CONSENT_WITHDRAWN_KEY, PipelineConfiguration.RQA_CODING_PLANS))
-    for rp in repeat_participations.values():
-        rp["% of Participants with Opt-Ins"] = \
-            round(rp["Number of Participants with Opt-Ins"] / total_participants * 100, 1)
-
-    # Export the participation frequency data to a csv
+    log.info("Computing repeat participations...")
     with open(f"{automated_analysis_output_dir}/repeat_participations.csv", "w") as f:
-        headers = ["Number of Episodes Participated In", "Number of Participants with Opt-Ins",
-                   "% of Participants with Opt-Ins"]
-        writer = csv.DictWriter(f, fieldnames=headers, lineterminator="\n")
-        writer.writeheader()
-
-        for row in repeat_participations.values():
-            writer.writerow(row)
+        repeat_participations.export_repeat_participations_csv(
+            individuals, CONSENT_WITHDRAWN_KEY,
+            coding_plans_to_analysis_configurations(PipelineConfiguration.RQA_CODING_PLANS),
+            f
+        )
 
     log.info(f'Computing repeat and new participation per show ...')
```

Comment:
How widely is this being used? This could be useful to standardise at some point.

Reply:
Mainly used in Wusc and Imaqal but its a relevant metric, will standardize and move it to core.